srml/authorship/src/lib.rs

```diff
@@ -0,0 +1,601 @@
+
+	let _ = uncles.drain(..prune_index);
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn on_initialize(now: T::BlockNumber) {
+			let uncle_generations = T::UncleGenerations::get();
+			let mut uncles = <Self as Store>::Uncles::get();
+
+			// prune uncles that are older than the allowed number of generations.
+			if uncle_generations <= now {
+				let minimum_height = now - uncle_generations;
+				prune_old_uncles(minimum_height, &mut uncles)
+			}
+
+			<Self as Store>::DidSetUncles::put(false);
+
+			T::EventHandler::note_author(Self::author());
+		}
+
+		fn on_finalize() {
+			// ensure we never go to trie with these values.
+			let _ = <Self as Store>::Author::take();
+			let _ = <Self as Store>::DidSetUncles::take();
+		}
+
+		/// Provide a set of uncles.
+		fn set_uncles(origin, new_uncles: Vec<T::Header>) -> DispatchResult {
+			ensure_none(origin)?;
+
+			if <Self as Store>::DidSetUncles::get() {
+				return Err("Uncles already set in block.");
+			}
+			<Self as Store>::DidSetUncles::put(true);
+
+			Self::verify_and_import_uncles(new_uncles)
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Fetch the author of the block.
+	///
+	/// This is safe to invoke `on_initialize` implementations as well as afterwards.
+	pub fn author() -> T::AccountId {
+		// Check the memoized storage value.
+		if let Some(author) = <Self as Store>::Author::get() {
+			return author;
+		}
+
+		let digest = <system::Module<T>>::digest();
+		let pre_runtime_digests = digest.logs.iter().filter_map(|d| d.as_pre_runtime());
+		if let Some(author) = T::FindAuthor::find_author(pre_runtime_digests) {
+			<Self as Store>::Author::put(&author);
+			author
+		} else {
+			Default::default()
+		}
+	}
+
+	fn verify_and_import_uncles(new_uncles: Vec<T::Header>) -> DispatchResult {
+		let now = <system::Module<T>>::block_number();
+
+		let (minimum_height, maximum_height) = {
+			let uncle_generations = T::UncleGenerations::get();
+			let min = if now >= uncle_generations {
+				now - uncle_generations
+			} else {
+				Zero::zero()
+			};
+
+			(min, now)
+		};
+
+		let mut uncles = <Self as Store>::Uncles::get();
+		uncles.push(UncleEntryItem::InclusionHeight(now));
+
+		let mut acc: <T::FilterUncle as FilterUncle<_, _>>::Accumulator = Default::default();
+
+		for uncle in new_uncles {
+			let hash = uncle.hash();
+
+			if uncle.number() < &One::one() {
+				return Err("uncle is genesis");
+			}
+
+			if uncle.number() > &maximum_height {
+				return Err("uncles too high in chain");
+			}
+
+			{
+				let parent_number = uncle.number().clone() - One::one();
+				let parent_hash = <system::Module<T>>::block_hash(&parent_number);
+				if &parent_hash != uncle.parent_hash() {
+					return Err("uncle parent not in chain");
+				}
+			}
+
+			if uncle.number() < &minimum_height {
+				return Err("uncle not recent enough to be included");
+			}
+
+			let duplicate = uncles.iter().find(|entry| match entry {
```

Comment:
The uncles list or a variant thereof is decent enough for tracking garbage to collect, but I agree we should shift the actual data into a map of `Hash -> bool`

Reply:
But then the main question is (given that in the code here we have the whole uncles list in contiguous memory) how much slower/faster the trie access to query the map is going to be vs. doing this iteration (for in practice, what is typically less than 50 elements).

We could also construct a `BTreeSet<Hash>` or something from the uncles list and then query that. It might be well faster than doing the few iterations. On the other hand, maybe not.

Reply:
If we need to load/store this every time, then an array is definitely faster.  On the other hand, if this is loaded once and then accessed many times, I would go with a `BTreeSet`.

Reply:
This also requires `T::Hash: Ord`, which isn't currently a bound on `system::Trait`. Let's consider this future optimization.